README.md
<p align="center"><a href="https://github.com/librenms/docker" target="_blank"><img height="128" src="https://raw.githubusercontent.com/librenms/docker/master/.github/docker-librenms.jpg"></a></p>

<p align="center">
  <a href="https://hub.docker.com/r/librenms/librenms/tags?page=1&ordering=last_updated"><img src="https://img.shields.io/github/v/tag/librenms/docker?label=version&style=flat-square" alt="Latest Version"></a>
  <a href="https://github.com/librenms/docker/actions?workflow=build"><img src="https://img.shields.io/github/actions/workflow/status/librenms/docker/build.yml?branch=master&label=build&logo=github&style=flat-square" alt="Build Status"></a>
  <a href="https://hub.docker.com/r/librenms/librenms/"><img src="https://img.shields.io/docker/stars/librenms/librenms.svg?style=flat-square&logo=docker" alt="Docker Stars"></a>
  <a href="https://hub.docker.com/r/librenms/librenms/"><img src="https://img.shields.io/docker/pulls/librenms/librenms.svg?style=flat-square&logo=docker" alt="Docker Pulls"></a>
  <br /><a href="https://github.com/sponsors/crazy-max"><img src="https://img.shields.io/badge/sponsor-crazy--max-181717.svg?logo=github&style=flat-square" alt="Become a sponsor"></a>
  <a href="https://www.paypal.me/crazyws"><img src="https://img.shields.io/badge/donate-paypal-00457c.svg?logo=paypal&style=flat-square" alt="Donate Paypal"></a>
</p>

## About

Docker image for [LibreNMS](https://www.librenms.org/), a fully featured network
monitoring system that provides a wealth of features and device support.

> **Note**
> 
> Want to be notified of new releases? Check out 🔔 [Diun (Docker Image Update Notifier)](https://github.com/crazy-max/diun)
> project!

___

* [Features](#features)
* [Demo](#demo)
* [Build locally](#build-locally)
* [Image](#image)
* [Environment variables](#environment-variables)
  * [General](#general)
  * [Redis](#redis)
  * [Dispatcher service](#dispatcher-service)
  * [Syslog-ng](#syslog-ng)
  * [Snmptrapd](#snmptrapd)
  * [Database](#database)
  * [Misc](#misc)
* [Volumes](#volumes)
* [Ports](#ports)
* [Usage](#usage)
  * [Docker Compose](#docker-compose)
  * [Command line](#command-line)
  * [First launch](#first-launch)
* [Upgrade](#upgrade)
* [Configuration Management](#configuration-management)
  * [Initial Configuration](#initial-configuration)
  * [Live Configuration](#live-configuration)
  * [Re-Apply YAML Config](#re-apply-yaml-config)
  * [Live Config](#live-config)
* [Notes](#notes)
  * [LNMS command](#lnms-command)
  * [Validate](#validate)
  * [Dispatcher service container](#dispatcher-service-container)
  * [Syslog-ng container](#syslog-ng-container)
  * [Snmptrapd container](#snmptrapd-container)
  * [Add a LibreNMS plugin](#add-a-librenms-plugin)
  * [Additional Monitoring plugins](#additional-monitoring-plugins)
  * [Custom alert templates](#custom-alert-templates)
* [Contributing](#contributing)
* [License](#license)

## Features

* Run as non-root user
* Multi-platform image
* [Dispatcher service](#dispatcher-service) as "sidecar" container
* Syslog-ng support through a ["sidecar" container](#syslog-ng)
* Snmp-trap support through a ["sidecar" container](#snmptrapd)
* Built-in LibreNMS [Weathermap plugin](https://docs.librenms.org/Extensions/Weathermap/)
* Ability to add custom Monitoring plugins
* Ability to add custom alert templates
* OPCache enabled to store precompiled script bytecode in shared memory
* [s6-overlay](https://github.com/just-containers/s6-overlay/) as process supervisor
* [Traefik](https://github.com/containous/traefik-library-image) as reverse proxy and creation/renewal of Let's Encrypt certificates (see [this template](examples/traefik))
* [Redis](https://github.com/docker-library/redis) image ready to use for better scalability
* [RRDcached](https://github.com/crazy-max/docker-rrdcached) image ready to use for data caching and graphs
* [msmtpd SMTP relay](https://github.com/crazy-max/docker-msmtpd) image to send emails
* [MariaDB](https://github.com/docker-library/mariadb) image as database instance

## Demo

[![Try in PWD](https://raw.githubusercontent.com/play-with-docker/stacks/master/assets/images/button.png)](https://labs.play-with-docker.com/?stack=https://raw.githubusercontent.com/librenms/docker/master/examples/pwd/librenms.yml)

## Build locally

```console
$ git clone https://github.com/librenms/docker.git docker-librenms
$ cd docker-librenms

# Build image and output to docker (default)
$ docker buildx bake

# Build multi-platform image
$ docker buildx bake image-all
```

## Image

Following platforms for this image are available:

```
$ docker run --rm mplatform/mquery librenms/librenms:latest
Image: librenms/librenms:latest
 * Manifest List: Yes
 * Supported platforms:
   - linux/amd64
   - linux/arm/v6
   - linux/arm/v7
   - linux/arm64
   - linux/386
   - linux/ppc64le
   - linux/s390x
```

## Environment variables

### General

* `TZ`: The timezone assigned to the container (default `UTC`)
* `PUID`: LibreNMS user id (default `1000`)
* `PGID`: LibreNMS group id (default `1000`)
* `MEMORY_LIMIT`: PHP memory limit (default `256M`)
* `MAX_INPUT_VARS`: PHP max input vars (default `1000`)
* `UPLOAD_MAX_SIZE`: Upload max size (default `16M`)
* `CLEAR_ENV`: Clear environment in FPM workers (default `yes`)
* `FPM_PM_MAX_CHILDREN`: FPM max Children (default: `15`)
* `FPM_PM_START_SERVERS`: FPM start servers (default: `2`)
* `FPM_PM_MIN_SPARE_SERVERS`: FPM min spare servers (default: `1`)
* `FPM_PM_MAX_SPARE_SERVERS`: FPM max spare servers (default: `6`)
* `OPCACHE_MEM_SIZE`: PHP OpCache memory consumption (default `128`)
* `LISTEN_IPV6`: Enable IPv6 for Nginx (default `true`)
* `REAL_IP_FROM`: Trusted addresses that are known to send correct replacement addresses (default `0.0.0.0/32`)
* `REAL_IP_HEADER`: Request header field whose value will be used to replace the client address (default `X-Forwarded-For`)
* `LOG_IP_VAR`: Use another variable to retrieve the remote IP address for access [log_format](http://nginx.org/en/docs/http/ngx_http_log_module.html#log_format) on Nginx. (default `remote_addr`)
* `SESSION_DRIVER`: [Driver to use for session storage](https://github.com/librenms/librenms/blob/master/config/session.php) (default `file`)
* `CACHE_DRIVER`: [Driver to use for cache and locks](https://github.com/librenms/librenms/blob/master/config/cache.php) (default `database`)
* `IGNORE_ERRORS`: If set to `1`, startup scripts won't exit if any failure is found. (default not defined).

### Redis

> **Note**
>
> Redis variables should be set on all containers and are required when running
> more than one dispatcher.

* `REDIS_HOST`: Redis host for poller synchronization
* `REDIS_SENTINEL`: Redis Sentinel host for high availability Redis cluster
* `REDIS_SENTINEL_SERVICE`: Redis Sentinel service name (default `librenms`)
* `REDIS_SCHEME`: Redis scheme (default `tcp`)
* `REDIS_PORT`: Redis port (default `6379`)
* `REDIS_PASSWORD`: Redis password
* `REDIS_DB`: Redis database (default `0`)
* `REDIS_CACHE_DB`: Redis cache database (default `1`)

### Dispatcher service

> **Warning**
>
> Only used if you enable and run a [sidecar dispatcher container](#dispatcher-service-container).

* `SIDECAR_DISPATCHER`: Set to `1` to enable sidecar dispatcher mode for this container (default `0`)
* `DISPATCHER_NODE_ID`: Unique node ID for your dispatcher service
* `DISPATCHER_ARGS`: Additional args to pass to the [dispatcher service](https://github.com/librenms/librenms/blob/master/librenms-service.py)

### Syslog-ng

> **Warning**
>
> Only used if you enable and run a [sidecar syslog-ng container](#syslog-ng-container).

* `SIDECAR_SYSLOGNG`: Set to `1` to enable sidecar syslog-ng mode for this container (default `0`)

### Snmptrapd

> **Warning**
>
> Only used if you enable and run a [sidecar snmptrapd container](#snmptrapd-container).

* `SIDECAR_SNMPTRAPD`: Set to `1` to enable sidecar snmptrapd mode for this container (default `0`)
* `SNMP_PROCESSING_TYPE`: Sets which type of processing (`log`, `execute`, and/or `net`) to use with the SNMP trap (default `log,execute,net`)
* `SNMP_USER`: Defines what username to authenticate with (default `librenms_user`)
* `SNMP_AUTH`: Defines what password to authenticate with (default `auth_pass` should not be used, but will work)
* `SNMP_PRIV`: Defines what password to encrypt packages with (default `priv_pass` should not be used, but will work)
* `SNMP_AUTH_PROTO`: Sets what protocol (`MD5`|`SHA`) to use for authentication (default `SHA`)
* `SNMP_PRIV_PROTO`: Sets what protocol (`DES`|`AES`) to use for encryption of packages (default `AES`)
* `SNMP_SECURITY_LEVEL`: Sets what security level (`noauth`|`priv`) to use (default `priv`)
* `SNMP_ENGINEID`: Defines what SNMP EngineID to use (default `1234567890`)
* `SNMP_DISABLE_AUTHORIZATION`: Will disable the above access control checks, and revert to the previous behaviour of accepting all incoming notifications. (default `yes`)

### Database

* `DB_HOST`: MySQL database hostname / IP address
* `DB_PORT`: MySQL database port (default `3306`)
* `DB_NAME`: MySQL database name (default `librenms`)
* `DB_USER`: MySQL user (default `librenms`)
* `DB_PASSWORD`: MySQL password (default `librenms`)
* `DB_TIMEOUT`: Time in seconds after which we stop trying to reach the MySQL server (useful for clusters, default `60`)

### Misc

* `LIBRENMS_BASE_URL`: URL of your LibreNMS instance (default `/`)
* `LIBRENMS_SNMP_COMMUNITY`: This container's SNMP v2c community string (default `librenmsdocker`)
* `LIBRENMS_WEATHERMAP`: Enable LibreNMS [Weathermap plugin](https://docs.librenms.org/Extensions/Weathermap/) (default `false`)
* `LIBRENMS_WEATHERMAP_SCHEDULE`: CRON expression format (default `*/5 * * * *`)
* `MEMCACHED_HOST`: Hostname / IP address of a Memcached server
* `MEMCACHED_PORT`: Port of the Memcached server (default `11211`)
* `RRDCACHED_SERVER`: RRDcached server (eg. `rrdcached:42217`)

## Volumes

* `/data`: Contains configuration, plugins, rrd database, logs, additional Monitoring plugins, additional syslog-ng config files

> **Warning**
>
> Note that the volume should be owned by the user/group with the specified
> `PUID` and `PGID`. If you don't give the volume correct permissions, the
> container may not start.

## Ports

* `8000`: HTTP port
* `514 514/udp`: Syslog ports (only used if you enable and run a [sidecar syslog-ng container](#syslog-ng-container))

CHANGELOG.md
# Changelog

## 23.5.0-r0 (2023/06/15)

* LibreNMS 23.5.0 (#357)
* Add extra SNMP Trap MIB dirs via env var (#355)

## 23.2.0-r0 (2023/03/24)

* LibreNMS 23.2.0 (#340)

## 23.1.1-r0 (2023/02/09)

* LibreNMS 23.1.1 (#335)

## 22.12.0-r0 (2022/12/30)

* LibreNMS 22.12.0 (#329)
* Alpine Linux 3.17 (#328)

## 22.11.0-r0 (2022/11/26)

* LibreNMS 22.11.0 (#324)
* Update to PHP 8.1 (#324)
* Add ability to configure php-fpm (#320)

## 22.10.0-r0 (2022/10/22)

* LibreNMS 22.10.0 (#318)
* Enable the First Run Wizard (#280)
* Add `openssh-client` package for the `check_by_ssh` plugin (#312)

## 22.9.0-r0 (2022/09/22)

* LibreNMS 22.9.0 (#314)

## 22.8.0-r1 (2022/08/23)

* Update weathermap plugin to `697ac3c` (#309)

## 22.8.0-r0 (2022/08/23)

* LibreNMS 22.8.0 (#306)

## 22.7.0-r2 (2022/08/14)

* Fix container restart state with Weathermap (#304)

## 22.7.0-r1 (2022/08/13)

* Remove Weathermap plugin while waiting for a fix upstream (#302)
* Fix symlink and perms for weathermap plugin configs folder (#301)
* Ensure `REDIS_SCHEME` is set with default value (#300)

## 22.7.0-r0 (2022/07/24)

* LibreNMS 22.7.0 (#295)
* Add workflow to test librenms master branch (#293)
* Always set `rrdtool` version (#292)
* Shallow clone to reduce build time (#291)
* YAML based config (#225)
* Move Redis env vars in `librenms.env` file (#290)
* Remove memcached and use Redis fully (#286)
* Verifying user and group in `lnms` script (#278)
* Update to PHP 8 and Alpine Linux 3.16 (#289)

## 22.6.0-r1 (2022/07/14)

* Add PHP GMP extension (#288)
* Ensure scripts exit early on error and set shellcheck (#283)
* Add `MAX_INPUT_VARS` env var (#271)
* Remove test data that is not used in production (#281)

## 22.6.0-r0 (2022/06/17)

* LibreNMS 22.6.0 (#275)

## 22.5.0-r0 (2022/05/28)

* LibreNMS 22.5.0 (#272)

## 22.4.1-r0 (2022/04/24)

* LibreNMS 22.4.1 (#261)

## 22.3.0-r0 (2022/03/20)

* LibreNMS 22.3.0 (#258)

## 22.2.2-r0 (2022/03/13)

* LibreNMS 22.2.2 (#257)

## 22.2.1-r0 (2022/02/20)

* LibreNMS 22.2.1 (#254)
* Alpine Linux 3.15 (#253)

## 22.2.0-r0 (2022/02/18)

* LibreNMS 22.2.0 (#252)

## 22.1.0-r0 (2022/01/26)

* LibreNMS 22.1.0 (#250)

## 21.12.1-r0 (2022/01/05)

* LibreNMS 21.12.1 (#247)
* Fix pip pkgs (#249)
* nginx: increase fastcgi buffer size (#246)

## 21.11.0-r0 (2021/11/14)

* LibreNMS 21.11.0 (#242)

## 21.10.2-r0 (2021/10/21)

* LibreNMS 21.10.2 (#239)

## 21.10.1-r0 (2021/10/19)

* LibreNMS 21.10.1 (#236)

## 21.10.0-r0 (2021/10/18)

* LibreNMS 21.10.0 (#235)

## 21.9.1-r0 (2021/09/28)

* LibreNMS 21.9.1 (#230)

## 21.9.0-r0 (2021/09/20)

* LibreNMS 21.9.0 (#224)

## 21.8.0-r0 (2021/09/01)

* LibreNMS 21.8.0 (#222)
* Alpine Linux 3.14 (#221)

## 21.7.0-r2 (2021/08/10)

* Misspelling of `SIDECAR_SNMPTRAPD` environment variable (#220)
* Fix `librenms_snmptrapd` service typo (#219)

## 21.7.0-r1 (2021/08/07)

* Add SNMP config (#216)

## 21.7.0-r0 (2021/08/01)

* LibreNMS 21.7.0 (#215)
* Create and enable a snmp trap handling container (#180)

## 21.6.0-r0 (2021/06/28)

* LibreNMS 21.6.0 (#210)

## 21.5.1-r0 (2021/05/20)

* LibreNMS 21.5.1 (#205)
* Move to `docker/metadata-action` (#199)
* Add `file` package (#195)

## 21.4.0-r0 (2021/04/18)

* LibreNMS 21.4.0 (#192)
* Update db service example (#193)
* Fix weathermap output perms (#191)
* Fix warnings (#190)
* All containers should connect to redis (#178)
* alpine-s6 3.13-2.2.0.3 (#189)

## 21.3.0-r0 (2021/03/21)

* LibreNMS 21.3.0
* Update syslog-ng max-connections (#179)

## 21.2.0-r3 (2021/03/18)

* Upstream Alpine update

## 21.2.0-r2 (2021/03/10)

* Handle LibreNMS plugins (#171)

## 21.2.0-r1 (2021/03/04)

* Switch to `yasu`

## 21.2.0-r0 (2021/02/16)

* LibreNMS 21.2.0 (#167)
* `s6-overlay` 2.2.0.3 (#162)
* Alpine Linux 3.13 (#162)

## 21.1.0-r2 (2021/02/16)

* Need to specify `rrdtool_version` (librenms/docker@ab027d7)

## 21.1.0-r1 (2021/02/13)

* Downgrade to `s6-overlay` 2.1.0.2 (#164)
* No need to specify `rrdtool_version` (#166)

## 21.1.0-r0 (2021/02/02)

* LibreNMS 21.1.0
* Switch to buildx bake
* Do not fail on permission issue
* Redis - Add scheme to allow TLS (#159)
* Add iputils and fix perms (#131)

## 1.70.1-RC2 (2020/12/10)

* Fix Redis for dispatcher

## 1.70.1-RC1 (2020/12/02)

* LibreNMS 1.70.1
* Add NET_ADMIN capability and fix fping6 (#140)

## 1.69-RC3 (2020/11/22)

* Fix tzdata issue with Alpine (#143)
* Add alert templates (#142)
* Add redis sentinel support (#141)

## 1.69-RC2 (2020/11/04)

* Do not set default value for `RRDCACHED_SERVER`

## 1.69-RC1 (2020/11/03)

* LibreNMS 1.69
* Update to Traefik v2
* Update PWD example (#135)
* Allow to clear env for FPM workers
* Use Docker meta action to handle tags and labels
* Replace `RRDCACHED_HOST` and `RRDCACHED_PORT` with `RRDCACHED_SERVER` env var

## 1.68-RC1 (2020/09/30)

* LibreNMS 1.68

## 1.67-RC2 (2020/09/04)

* Seed through artisan (#122)

## 1.67-RC1 (2020/09/03)

* LibreNMS 1.67
* Now based on [Alpine Linux 3.12 with s6 overlay](https://github.com/crazy-max/docker-alpine-s6/)
* Missing migration seed flag (#122)

## 1.66-RC2 (2020/08/28)

* Fix DB connection for dispatcher service (#108 #118 #119)
* Run maintenance task through a dedicated process (#105)
* Add `DISPATCHER_ARGS` env var


entrypoint.sh
#!/bin/bash

function runas_librenms() {
  su - librenms -s /bin/sh -c "$1"
}

TZ=${TZ:-UTC}
PUID=${PUID:-1000}
PGID=${PGID:-1000}

MEMORY_LIMIT=${MEMORY_LIMIT:-256M}
UPLOAD_MAX_SIZE=${UPLOAD_MAX_SIZE:-16M}
OPCACHE_MEM_SIZE=${OPCACHE_MEM_SIZE:-128}
REAL_IP_FROM=${REAL_IP_FROM:-"0.0.0.0/32"}
REAL_IP_HEADER=${REAL_IP_HEADER:-"X-Forwarded-For"}
LOG_IP_VAR=${LOG_IP_VAR:-remote_addr}

LIBRENMS_POLLER_THREADS=${LIBRENMS_POLLER_THREADS:-16}
LIBRENMS_POLLER_INTERVAL=${LIBRENMS_POLLER_INTERVAL:-5}

LIBRENMS_DISTRIBUTED_POLLER_ENABLE=${LIBRENMS_DISTRIBUTED_POLLER_ENABLE:-false}
LIBRENMS_DISTRIBUTED_POLLER_NAME=${LIBRENMS_DISTRIBUTED_POLLER_NAME:-$(hostname -f)}
LIBRENMS_DISTRIBUTED_POLLER_GROUP=${LIBRENMS_DISTRIBUTED_POLLER_GROUP:-'0'}
LIBRENMS_DISTRIBUTED_POLLER_MEMCACHED_HOST=${LIBRENMS_DISTRIBUTED_POLLER_MEMCACHED_HOST:-${MEMCACHED_HOST}}
LIBRENMS_DISTRIBUTED_POLLER_MEMCACHED_PORT=${LIBRENMS_DISTRIBUTED_POLLER_MEMCACHED_PORT:-${MEMCACHED_PORT}}

SIDECAR_CRON=${SIDECAR_CRON:-0}
LIBRENMS_CRON_DISCOVERY_ENABLE=${LIBRENMS_CRON_DISCOVERY_ENABLE:-true}
LIBRENMS_CRON_DAILY_ENABLE=${LIBRENMS_CRON_DAILY_ENABLE:-true}
LIBRENMS_CRON_ALERTS_ENABLE=${LIBRENMS_CRON_ALERTS_ENABLE:-true}
LIBRENMS_CRON_BILLING_ENABLE=${LIBRENMS_CRON_BILLING_ENABLE:-true}
LIBRENMS_CRON_BILLING_CALCULATE_ENABLE=${LIBRENMS_CRON_BILLING_CALCULATE_ENABLE:-true}
LIBRENMS_CRON_CHECK_SERVICES_ENABLE=${LIBRENMS_CRON_CHECK_SERVICES_ENABLE:-true}
LIBRENMS_CRON_POLLER_ENABLE=${LIBRENMS_CRON_POLLER_ENABLE:-true}

SIDECAR_SYSLOGNG=${SIDECAR_SYSLOGNG:-0}

DB_PORT=${DB_PORT:-3306}
DB_NAME=${DB_NAME:-librenms}
DB_USER=${DB_USER:-librenms}
DB_TIMEOUT=${DB_TIMEOUT:-30}

MEMCACHED_PORT=${MEMCACHED_PORT:-11211}
RRDCACHED_PORT=${RRDCACHED_PORT:-42217}

# From https://github.com/docker-library/mariadb/blob/master/docker-entrypoint.sh#L21-L41
# usage: file_env VAR [DEFAULT]
#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
# (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
#  "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
file_env() {
  local var="$1"
  local fileVar="${var}_FILE"
  local def="${2:-}"
  if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then
    echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
    exit 1
  fi
  local val="$def"
  if [ "${!var:-}" ]; then
    val="${!var}"
  elif [ "${!fileVar:-}" ]; then
    val="$(< "${!fileVar}")"
  fi
  export "$var"="$val"
  unset "$fileVar"
}

# Timezone
echo "Setting timezone to ${TZ}..."
ln -snf /usr/share/zoneinfo/${TZ} /etc/localtime
echo ${TZ} > /etc/timezone
sed -i -e "s|date\.timezone.*|date\.timezone = ${TZ}|" /etc/php7/php.ini \

# Change librenms UID / GID
echo "Checking if librenms UID / GID has changed..."
if [ -n "${PGID}" ] && [ "${PGID}" != "`id -g librenms`" ]; then
  sed -i -e "s/^librenms:\([^:]*\):[0-9]*/librenms:\1:${PGID}/" /etc/group
  sed -i -e "s/^librenms:\([^:]*\):\([0-9]*\):[0-9]*/librenms:\1:\2:${PGID}/" /etc/passwd
fi
if [ -n "${PUID}" ] && [ "${PUID}" != "`id -u librenms`" ]; then
  sed -i -e "s/^librenms:\([^:]*\):[0-9]*:\([0-9]*\)/librenms:\1:${PUID}:\2/" /etc/passwd
fi

# PHP
echo "Setting PHP-FPM configuration..."
sed -e "s/@MEMORY_LIMIT@/$MEMORY_LIMIT/g" \
  -e "s/@UPLOAD_MAX_SIZE@/$UPLOAD_MAX_SIZE/g" \
  /tpls/etc/php7/php-fpm.d/www.conf > /etc/php7/php-fpm.d/www.conf
echo "Setting PHP INI configuration..."
sed -i -e "s|memory_limit.*|memory_limit = ${MEMORY_LIMIT}|" /etc/php7/php.ini

# OpCache
echo "Setting OpCache configuration..."
sed -e "s/@OPCACHE_MEM_SIZE@/$OPCACHE_MEM_SIZE/g" \
  /tpls/etc/php7/conf.d/opcache.ini > /etc/php7/conf.d/opcache.ini

# Nginx
echo "Setting Nginx configuration..."
sed -e "s#@UPLOAD_MAX_SIZE@#$UPLOAD_MAX_SIZE#g" \
  -e "s#@REAL_IP_FROM@#$REAL_IP_FROM#g" \
  -e "s#@REAL_IP_HEADER@#$REAL_IP_HEADER#g" \
  -e "s#@LOG_IP_VAR@#$LOG_IP_VAR#g" \
  /tpls/etc/nginx/nginx.conf > /etc/nginx/nginx.conf

# SNMP
echo "Updating SNMP community..."
file_env 'LIBRENMS_SNMP_COMMUNITY' 'librenmsdocker'
sed -i -e "s/RANDOMSTRINGGOESHERE/${LIBRENMS_SNMP_COMMUNITY}/" /etc/snmp/snmpd.conf

# Init files and folders
echo "Initializing LibreNMS files / folders..."
mkdir -p ${DATA_PATH}/config \
  ${DATA_PATH}/logs \
  ${DATA_PATH}/monitoring-plugins \
  ${DATA_PATH}/rrd
rm -f ${LIBRENMS_PATH}/config.d/*

echo "Setting LibreNMS configuration..."

# Config : Directories
  cat > ${LIBRENMS_PATH}/config.d/directories.php <<EOL
<?php
\$config['install_dir'] = '${LIBRENMS_PATH}';
\$config['log_dir'] = '${DATA_PATH}/logs';
\$config['rrd_dir'] = '${DATA_PATH}/rrd';
EOL

# Config : Database
if [ -z "$DB_HOST" ]; then
  >&2 echo "ERROR: DB_HOST must be defined"
  exit 1
fi
file_env 'DB_PASSWORD'
if [ -z "$DB_PASSWORD" ]; then
  >&2 echo "ERROR: Either DB_PASSWORD or DB_PASSWORD_FILE must be defined"
  exit 1
fi
cat > ${LIBRENMS_PATH}/config.d/database.php <<EOL
<?php
\$config['db_host'] = '${DB_HOST}';
\$config['db_port'] = ${DB_PORT};
\$config['db_user'] = '${DB_USER}';
\$config['db_pass'] = '${DB_PASSWORD}';
\$config['db_name'] = '${DB_NAME}';
EOL
dbcmd="mysql -h ${DB_HOST} -P ${DB_PORT} -u "${DB_USER}" "-p${DB_PASSWORD}""
unset DB_PASSWORD

# Config : User
cat > ${LIBRENMS_PATH}/config.d/user.php <<EOL
<?php
\$config['user'] = "librenms";
EOL

# Config : Fping
echo "/usr/sbin/fping -6 \$@" > /usr/sbin/fping6
chmod +x /usr/sbin/fping6
cat > ${LIBRENMS_PATH}/config.d/fping.php <<EOL
<?php
\$config['fping'] = "/usr/sbin/fping";
\$config['fping6'] = "/usr/sbin/fping6";
EOL

# Config : ipmitool
cat > ${LIBRENMS_PATH}/config.d/ipmitool.php <<EOL
<?php
\$config['ipmitool'] = "/usr/sbin/ipmitool";
EOL

# Config : Disable autoupdate
cat > ${LIBRENMS_PATH}/config.d/autoupdate.php <<EOL
<?php
\$config['update'] = 0;
EOL

# Config : Services
cat > ${LIBRENMS_PATH}/config.d/services.php <<EOL
<?php
\$config['show_services'] = 1;
\$config['nagios_plugins'] = "/usr/lib/monitoring-plugins";
EOL

# Config : Memcached
if [ ! -z "${MEMCACHED_HOST}" ]; then
    cat > ${LIBRENMS_PATH}/config.d/memcached.php <<EOL
<?php
\$config['memcached']['enable'] = true;
\$config['memcached']['host'] = '${MEMCACHED_HOST}';
\$config['memcached']['port'] = ${MEMCACHED_PORT};
EOL
fi

# Config : RRDcached
if [ ! -z "${RRDCACHED_HOST}" ]; then
    cat > ${LIBRENMS_PATH}/config.d/rrdcached.php <<EOL
<?php
\$config['rrdcached'] = "${RRDCACHED_HOST}:${RRDCACHED_PORT}";
\$config['rrdtool_version'] = '1.7.0';
EOL
fi

# Config : Ditributed poller
if [ ! -z "${LIBRENMS_DISTRIBUTED_POLLER_MEMCACHED_HOST}" -a ! -z "${RRDCACHED_HOST}" -a $LIBRENMS_DISTRIBUTED_POLLER_ENABLE = true ]; then
    cat > ${LIBRENMS_PATH}/config.d/distributed_poller.php <<EOL
<?php
\$config['distributed_poller'] = true;
\$config['distributed_poller_name'] = '${LIBRENMS_DISTRIBUTED_POLLER_NAME}';
\$config['distributed_poller_group'] = '${LIBRENMS_DISTRIBUTED_POLLER_GROUP}';
\$config['distributed_poller_memcached_host'] = '${LIBRENMS_DISTRIBUTED_POLLER_MEMCACHED_HOST}';
\$config['distributed_poller_memcached_port'] = ${LIBRENMS_DISTRIBUTED_POLLER_MEMCACHED_PORT};
EOL
fi

 # Fix perms
echo "Fixing permissions..."
chown librenms. ${DATA_PATH}/config \
  ${DATA_PATH}/logs \
  ${DATA_PATH}/monitoring-plugins \
  ${DATA_PATH}/rrd
chown -R librenms. ${LIBRENMS_PATH}/config.d \
  ${LIBRENMS_PATH}/bootstrap \
  ${LIBRENMS_PATH}/storage
chmod ug+rw ${DATA_PATH}/logs \
  ${DATA_PATH}/rrd \
  ${LIBRENMS_PATH}/bootstrap/cache \
  ${LIBRENMS_PATH}/storage \
  ${LIBRENMS_PATH}/storage/framework/*

# Check additional Monitoring plugins
echo "Checking additional Monitoring plugins..."
mon_plugins=$(ls -l ${DATA_PATH}/monitoring-plugins | egrep '^-' | awk '{print $9}')
for mon_plugin in ${mon_plugins}; do
  if [ -f "/usr/lib/monitoring-plugins/${mon_plugin}" ]; then
    echo "  WARNING: Official Monitoring plugin ${mon_plugin} cannot be overriden. Skipping..."
    continue
  fi
  if [[ ${mon_plugin} != check_* ]]; then
    echo "  WARNING: Monitoring plugin filename ${mon_plugin} invalid. It must start with 'check_'. Skipping..."
    continue
  fi
  if [[ ! -x "${DATA_PATH}/monitoring-plugins/${mon_plugin}" ]]; then
    echo "  WARNING: Monitoring plugin file ${mon_plugin} has to be executable. Skipping..."
    continue
  fi
  echo "  Adding ${mon_plugin} Monitoring plugin"
  ln -sf ${DATA_PATH}/monitoring-plugins/${mon_plugin} /usr/lib/monitoring-plugins/${mon_plugin}
done

# Sidecar cron container
if [ "$SIDECAR_CRON" = "1" ]; then
  echo ">>"
  echo ">> Sidecar cron container detected"
  echo ">>"

  # Init
  rm /etc/supervisord/nginx.conf /etc/supervisord/php.conf /etc/supervisord/snmpd.conf /etc/supervisord/syslog-ng.conf
  if [ -z "$CRONTAB_PATH" ]; then
    >&2 echo "ERROR: CRONTAB_PATH must be defined"
    exit 1
  fi

  rm -rf ${CRONTAB_PATH}
  mkdir -m 0644 -p ${CRONTAB_PATH}
  touch ${CRONTAB_PATH}/librenms

  # Add crontab
  cat ${LIBRENMS_PATH}/librenms.nonroot.cron > ${CRONTAB_PATH}/librenms
  sed -i -e "s/ librenms //" ${CRONTAB_PATH}/librenms

  if [ $LIBRENMS_CRON_DISCOVERY_ENABLE != true ]; then
    sed -i "/discovery.php/d" ${CRONTAB_PATH}/librenms
  fi

  if [ $LIBRENMS_CRON_DAILY_ENABLE != true ]; then
    sed -i "/daily.sh/d" ${CRONTAB_PATH}/librenms
  fi

  if [ $LIBRENMS_CRON_ALERTS_ENABLE != true ]; then
    sed -i "/alerts.php/d" ${CRONTAB_PATH}/librenms
  fi

  if [ $LIBRENMS_CRON_BILLING_ENABLE != true ]; then
    sed -i "/poll-billing.php/d" ${CRONTAB_PATH}/librenms
  fi

  if [ $LIBRENMS_CRON_BILLING_CALCULATE_ENABLE != true ]; then
    sed -i "/billing-calculate.php/d" ${CRONTAB_PATH}/librenms
  fi

  if [ $LIBRENMS_CRON_CHECK_SERVICES_ENABLE != true ]; then
    sed -i "/check-services.php/d" ${CRONTAB_PATH}/librenms
  fi

  sed -i "/poller-wrapper.py/d" ${CRONTAB_PATH}/librenms
  if [ $LIBRENMS_CRON_POLLER_ENABLE = true ]; then
    cat >> ${CRONTAB_PATH}/librenms <<EOL
*/${LIBRENMS_POLLER_INTERVAL}  *    * * *     /opt/librenms/cronic /opt/librenms/poller-wrapper.py ${LIBRENMS_POLLER_THREADS}
EOL
  fi

  # Fix crontab perms
  echo "Fixing crontab permissions..."
  chmod -R 0644 ${CRONTAB_PATH}
elif [ "$SIDECAR_SYSLOGNG" = "1" ]; then
  echo ">>"
  echo ">> Sidecar syslog-ng container detected"
  echo ">>"

  # Init
  rm /etc/supervisord/cron.conf /etc/supervisord/nginx.conf /etc/supervisord/php.conf /etc/supervisord/snmpd.conf
  mkdir -p ${DATA_PATH}/syslog-ng /run/syslog-ng
  chown librenms. ${DATA_PATH}/syslog-ng
  chown -R librenms. /run/syslog-ng
else
  # Init
  rm /etc/supervisord/cron.conf /etc/supervisord/syslog-ng.conf

  echo "Waiting ${DB_TIMEOUT}s for database to be ready..."
  counter=1
  while ! ${dbcmd} -e "show databases;" > /dev/null 2>&1; do
      sleep 1
      counter=$((counter + 1))
      if [ ${counter} -gt ${DB_TIMEOUT} ]; then
          >&2 echo "ERROR: Failed to connect to database on $DB_HOST"
          exit 1
      fi;
  done
  echo "Database ready!"

  counttables=$(echo 'SHOW TABLES' | ${dbcmd} "$DB_NAME" | wc -l)

  echo "Updating database schema..."
  runas_librenms "php build-base.php"

  if [ "${counttables}" -eq "0" ]; then
    echo "Creating admin user..."
    runas_librenms "php adduser.php librenms librenms 10 librenms@librenms.docker"
  fi
fi

exec "$@"
